CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+--Panel3D: 3D GUI panels for VR
+--written by 0xBAADF00D
+--revised/refactored 5/11/16
+
+local UserInputService = game:GetService("UserInputService")
+local RunService = game:GetService("RunService")
+local GuiService = game:GetService("GuiService")
+local CoreGui = game:GetService("CoreGui")
+local RobloxGui = CoreGui:WaitForChild("RobloxGui")
+
+
+--Util
+local Util = {}
+do
+	function Util.Create(instanceType)
+		return function(data)
+			local obj = Instance.new(instanceType)
+			for k, v in pairs(data) do
+				if type(k) == 'number' then
+					v.Parent = obj
+				else
+					obj[k] = v
+				end
+			end
+			return obj
+		end
+	end
+
+	-- RayPlaneIntersection (shortened)
+	-- http://www.siggraph.org/education/materials/HyperGraph/raytrace/rayplane_intersection.htm
+	function Util.RayPlaneIntersection(ray, planeNormal, pointOnPlane)
+		planeNormal = planeNormal.unit
+		ray = ray.Unit
+
+		local Vd = planeNormal:Dot(ray.Direction)
+		if Vd == 0 then -- parallel, no intersection
+			return nil
+		end
+
+		local V0 = planeNormal:Dot(pointOnPlane - ray.Origin)
+		local t = V0 / Vd
+		if t < 0 then --plane is behind ray origin, and thus there is no intersection
+			return nil
+		end
+		
+		return ray.Origin + ray.Direction * t
+	end
+end
+--End of Util
+
+
+--Panel3D State variables
+local coreScriptMode = true
```

Comment:
This seems like a constant, we like to name all of the constants with the following style: `CORE_SCRIPT_MODE`